### PR TITLE
Update redundant arg matcher message to include example uses

### PR DESCRIPTION
**Pull request by dtchepak**
Not sure if the message is too wordy now, but I wanted to try to add some more information on how to fix the problem and some examples of misusing the arg matchers. Feel free to reject/reword.

### CODE REVIEW

File: src/NSubstitute/Exceptions/RedundantArgumentMatcherException.cs [hunk 1]
```diff
@@ -1,11 +1,50 @@
...
+            NewLine +
+            "This is often caused by using an argument spec with a call to a member NSubstitute does not handle" +
+            " (such as a non-virtual member or a call to an instance which is not a substitute), or for a purpose" +
+            " other than specifying a call (such as using an arg spec as a return value). For example:" +
+            NewLine +
+            NewLine +
+            "    var sub = Substitute.For<SomeClass>();" + NewLine +
+            "    var realType = new MyRealType(sub);" + NewLine +
+            "    // INCORRECT, arg spec used on realType, not a substitute:" + NewLine +
+            "    realType.SomeMethod(Arg.Any<int>());" + NewLine +
```
Comment (zvirja) [inline src/NSubstitute/Exceptions/RedundantArgumentMatcherException.cs:23]: Have you forgot to use `Returns()` extension method there? Otherwise, the sample is very non-natural and it's totally unclear why you'd write like that 😕

File: src/NSubstitute/Exceptions/AmbiguousArgumentsException.cs [hunk 2]
```diff
@@ -6,6 +6,12 @@ namespace NSubstitute.Exceptions
 {
     public class AmbiguousArgumentsException : SubstituteException
     {
+        private static readonly string BaseMessage =
```
Comment (zvirja) [inline src/NSubstitute/Exceptions/AmbiguousArgumentsException.cs]: Do you see any value in splitting the message constant and the actual string with arguments? I thought that string interpolation make it look better, so the variable and static parts are bundled seamlessly and you see the result at once, without switching between constant and concatenation..

File: src/NSubstitute/Exceptions/RedundantArgumentMatcherException.cs [hunk 3]
```diff
@@ -20,19 +59,13 @@ public RedundantArgumentMatcherException(string message) : base(message)
...
-                $" of the unexpected NSubstitute usage and most likely your test code behaves not as intended.{Environment.NewLine}" +
-                $"Please use the diagnostics information below to detect and fix the defective places.{Environment.NewLine}" +
-                $"{Environment.NewLine}" +
-                $"Remaining (non-bound) argument specifications:{Environment.NewLine}" +
-                $"{string.Join(Environment.NewLine, remainingSpecifications)}{Environment.NewLine}" +
-                $"{Environment.NewLine}" +
-                $"All argument specifications:{Environment.NewLine}" +
-                $"{string.Join(Environment.NewLine, allSpecifications)}";
+                BaseMessage +
+                $"Remaining (non-bound) argument specifications:{ItemSeparator}" +
```
Comment (zvirja) [inline src/NSubstitute/Exceptions/RedundantArgumentMatcherException.cs]: Personally, I feel that interpolation would work better here - we are splitting message parts across the whole file 😖

**Comment by zvirja** [comment]
@dtchepak Thanks for your improvements! It makes it look much better with your samples. I also love the long messages, when exception reason is well described. I've left a few notices, let me know if you wish to keep as is 😉

File: src/NSubstitute/Exceptions/RedundantArgumentMatcherException.cs (hunk 1 above)
Comment (dtchepak) [reply src/NSubstitute/Exceptions/RedundantArgumentMatcherException.cs:23]: Yes, I forgot to use `.Returns()` 😆 Sorry!

File: src/NSubstitute/Exceptions/AmbiguousArgumentsException.cs (hunk 2 above)
Comment (dtchepak) [reply src/NSubstitute/Exceptions/AmbiguousArgumentsException.cs]: As the message got longer I thought it was nice to separate out the dynamic and the static part.  This is subjective so I don't mind either way. :) 

Does the string interpolation stuff get compiled down to a constant where possible (the parts like `$"Some text{Environment.NewLine}" +$"More text{Environment.NewLine}"`)? Or is it dynamically created each time? If dynamic it might be better to keep it separate so it only creates the constant part of the string once.

File: src/NSubstitute/Exceptions/RedundantArgumentMatcherException.cs (hunk 3 above)
Comment (dtchepak) [reply src/NSubstitute/Exceptions/RedundantArgumentMatcherException.cs]: See https://github.com/zvirja/NSubstitute/pull/4#discussion_r170766208. I don't feel strongly about it. :)

File: src/NSubstitute/Exceptions/AmbiguousArgumentsException.cs (hunk 2 above)
Comment (zvirja) [reply src/NSubstitute/Exceptions/AmbiguousArgumentsException.cs]: >  Or is it dynamically created each time?

I've just checked. In this particular case it performs a concatenation of the multiple strings, formatted using the `string.Format()` call. However, I don't see any reason to worry about that. This is not a hot place, moreover - it's an exceptional case. Therefore, it doesn't matter how much strings we need to concatenate to raise an exception - those things are supposed to never happen 😅

In this case I'd prefer to focus on the code readability and, personally, I feel that interpolation is read better.

Let's probably also see @alexandrnikitin opinion, so it's clear for me what to do 😉

Comment (alexandrnikitin) [reply src/NSubstitute/Exceptions/AmbiguousArgumentsException.cs]: Either way works for me. I would go with string interpolation. You can go with multi line string interpolation using `$@" ...  "` It will use only one call to `string.Format` but I wouldn't worry about allocations here.

**Comment by zvirja** [comment]
I've merged all the suggested changes to my initial PR. 
Decided to proceed with interpolation as Alexandr doesn't mind 💃
